README.md
GitHubReplicator
================

# Description

GitHubReplicator is an application that replicates its own source in a new
GitHub repository when you navigate to its URL. 

When you navigate to its URL, it gets sufficient permissions to create a public
repository (which does not include access to private repositories). It then
forks the original repository at
<https://github.com/lkingsford/github_replicator.git> into a new fork in the
users GitHub profile. Finally, it redirects the browser to the new repository.

# Installation

## Required (assumed already set up):
- Debian 10 instance:

    - With root credentials
    - (Optionally) With DNS configured with A Record to that Debian instance

- GitHub Account in good standing

- Log into instance as root

## Register GitHub App

- Log on to GitHub using your account. 

- Navigate to <https://github.com/settings/applications/new>

- Set the Application Name as required. Set the homepage URL to the URL where
  the application is stored. Set the Authorization URL to 
  `[HOST]/github_callback` where `[HOST]` is the Hostname or IP address where
  the application is stored.

- Click 'Register Application'

- Store the 'Client ID' and the 'Client Secret' for use during the Server set up

## Server set up

- Install required packages

```
apt install nginx python3-pip python-dev certbot python-certbot-nginx git -y
pip3 install virtualenv
```

- Clone repository

> This is an active URL for the GitHubReplicator. Given the nature of the
> application, feel free to use a URL from a replicant repository.

```
git clone https://github.com/lkingsford/github_replicator.git
cd github_replicator
```

- Create and activate virtual environment

```
virtualenv -ppython3 env
. env/bin/activate
```

- Install required Python packages

```
pip install flask gunicorn python-dotenv requests
```

> The following steps are to set up the environment. Note that a .env file
> could be used (as shown here), or environment variables could be set
> appropriately for the platform (such as via the AWS Console or amending the
> Virtual Environment `activate` script

- Create the `.env` file by copying the `dotenv_template.txt` to `.env`

```
cp dotenv_template.txt .env
```

- Modify `.env` with your preferred text editor to set each value. A value is
  set by placing is directly after the appropriate `=` without a space,
  followed by a new line

    - `HOST` is the hostname of the server.
    - `GITHUB_CLIENT_ID` is the client ID shown by GitHub when creating the
       application.
    - `GITHUB_CLIENT_SECRET` is the client secret shown by GitHub when creating
       the application.
    - `STATE_KEY` is a cryptographically random string of numbers, and
       uppercase and lowercase letters, unique to the instance. It is
       recommended to generate it with KeePass or equivalent.
    - `LOG_FILENAME` is a location to store the log file. It is recommended to
       set it to a path inside the install folder - for instance,
       `/root/github_replicator/replicator.log`

- Use your preferred text editor to change the string SERVERNAME to the servername in `nginx_site_template.txt`

- Create the nginx site from the template

```
cp ngnix_site_template.txt /etc/nginx/sites-available/replicator
```

- Create the Gunicorn service

```
cp service_template /etc/systemd/system/replicator.service
service replicator start
```

  hostname of the server

- Make the site active

```
ln /etc/nginx/sites-available/replicator /etc/nginx/sites-enabled/replicator
```

- Create and install an SSL certificate

```
certbot --nginx
```
and follow the instructions ensuring you select to redirect all HTTP traffic to HTTPS

The site should now be active.

history
    1  sudo apt update
    2  apt install nginx
    3  apt install nginx -y
    4  sudo apt install neovim
    5  cd /etc/nginx/
    6  ls
    7  cd sites-
    8  cd sites-available/
    9  ls
   10  cd def
   11  cd default
   12  nvim default 
   13  cd ..
   14  ls
   15  cd sites-enabled/
   16  ls
   17  rm default 
   18  cd ..
   19  cd sites-available/
   20  ls
   21  nvim default 
   22  cp default drtest.nerdygentleman.com
   23  nvim drtest.nerdygentleman.com 
   24  cd ~
   25  ls
   26  cd /var/www
   27  ls
   28  mkdir drtest
   29  cd drtest
   30  cd ..
   31  rm -r drtest
   32  mkdir drtest.nerdygentleman.com
   33  cd drtest.nerdygentleman.com/
   34  python3
   35  pip
   36  LS
   37  ls
   38  cd drtest.nerdygentleman.com/
   39  ls
   40  ls -a
   41  nvim app.py 
   42  nvim app.py 
   43  rm "/root/.local/share/nvim/swap//%root%drtest.nerdygentleman.com%app.py.swp"
   44  gunicorn --bind 0.0.0.0:5000  wsgi:app
   45  . env/bin/activate
   46  gunicorn --bind 0.0.0.0:5000  wsgi:app
   47  pip3 install flask gunicorn
   48  gunicorn --bind 0.0.0.0:5000 wsgi:app
   49  fg
   50  nvim app.py 
   51  gunicorn --bind 0.0.0.0:5000 wsgi:app
   52  nvim app.py 
   53  ls
   54  nvim wsgi.py 
   55  pip3 install flask
   56  nvim app.py 
   57  pip3 install flask
   58  gunicorn --bind 0.0.0.0:5000 wsgi:app
   59  sudo apt-get install python-pip python-dev
   60  mkdir drtest.nerdygentleman.com
   61  cd drtest.nerdygentleman.com/
   62  virtualenv env -ppython3
   63  venv
   64  sudo apt-get install python-virtualenv
   65  sudo apt-get install python3-virtualenv
   66  sudo apt-get install python3-pip
   67  virtualenv -ppython3 env
   68  python3-0virtualenv -ppython3 env
   69  python3-virtualenv -ppython3 en
   70  venv
   71  virtualenv
   72  sudo apt-get install python3-virtualenv
   73  pip
   74  pip3
   75  pip3 install virtualenv
   76  virtualenv
   77  /usr/local/bin/virtualenv
   78  sudo apt uninstall python3-virtualenv
   79  sudo apt remove python3-virtualenv
   80  pip install virtualenv
   81  virtualenv
   82  pip3 install virtualenv
   83  virtualenv -ppython3
   84  virtualenv -ppython3 env
   85  nvim app.py
   86  makeuser
   87  adduser
   88  adduser drtest
   89  sudo adduser --system --disabled-password drtest
   90  sudo removeuser drtest
   91  sudo deluser drtest
   92  sudo adduser --system --disabled-password drtest
   93  rm -r /home/drtest
   94  deluser drtest
   95  sudo adduser --system --disabled-password drtest
   96  cp -r drtest.nerdygentleman.com/ /home/drtest
   97  cd /home/drtest
   98  chown -r * drtest
   99  chown drtest *
  100  ls
  101  ls -a
  102  ls -l
  103  cd drtest.nerdygentleman.com/
  104  ls
  105  ls -l
  106  chown drtest * -R
  107  ls -l
  108  ls
  109  nvim app.py
  110  nvim /etc/systemd/system/drtest.service
  111  ls
  112  mkdir src
  113  rm -r __pycache__/
  114  mv app.py wsgi.py src
  115  ls
  116  fg
  117  service drtest start
  118  mail
  119  systemctl start drtest
  120  systemctl enable drtest
  121  ps --aux
  122  ps
  123  ps aux
  124  ps aux | grep python
  125  ps aux | grep gunicorn
  126  sudo su drtest
  127  su drtest
  128  sudo -u drtest mail
  129  nvim /etc/systemd/system/drtest.service
  130  service
  131  service --status-all
  132  systemctl start drtest
  133  systemctl start drtest --vv
  134  systemctl start drtest --v
  135  systemctl start drtest
  136  service --status-all
  137  service drtest
  138  service drtest start
  139  service drtest.service start
  140  service --status-all
  141  service drtest status
  142  cd env/bin/
  143  ls -a
  144  ls -l
  145  ls
  146  ls -a
  147  sudo -u drtest ./gunicorn
  148  sudo -u drtest bash
  149  ls
  150  ls /etc/systemd/system/
  151  ls /etc/systemd/system/ -l
  152  chmod 755 /etc/systemd/system/drtest.service 
  153  service stasrt drtest
  154  service drtest start
  155  service drtest status
  156  su drtest
  157  sudo -u drtest bash
  158  service drtest start
  159  service drtest status
  160  service drtest status
  161  service drtest status
  162  nvim /etc/systemd/system/drtest.service 
  163  service drtest start
  164  systemctl daemon-reload
  165  service drtest start
  166  service drtest statuus
  167  service drtest status
  168  service drtest status
  169  fg
  170  nvim /etc/systemd/system/drtest.service 
  171  systemctl daemon-reload
  172  service drtest start
  173  service drtest status
  174  service drtest status
  175  nvim /etc/nginx/sites-available/drtest.nerdygentleman.com 
  176  ln -s /etc/nginx/sites-available/drtest.nerdygentleman.com /etc/nginx/sites-enabled
  177  sudo systemctl restart nginx
  178  systemctl status nginx.service
  179  nvim /etc/nginx/sites-available/drtest.nerdygentleman.com 
  180  sudo systemctl restart nginx
  181  systemctl status nginx.service
  182  service nginx stop
  183  sudo apt install certbot
  184  certbot certonly --standalone
  185  sudo service start nginx
  186  sudo service nginx start
  187  sudo certbot --nginx
  188  apt install python-certbot-nginx
  189  sudo certbot --nginx
  190  nvim /etc/nginx/sites-available/drtest.nerdygentleman.com 
  191  sudo su drtest
  192  sudo -u drtest bash
  193  service drtest restart
  194  service drtest status
  195  sudo su drtest
  196  sudo -u drtest bash
  197  service drtest restart
  198  service drtest status
  199  service drtest status
  200  nvim /home/drtest/drtest.nerdygentleman.com/
  201  service drtest status
  202  service drtest restart
  203  service drtest status
  204  service drtest status
  205  nvim /home/drtest/drtest.nerdygentleman.com/src/app.py 
  206  service drtest restart
  207  service drtest status
  208  nvim /home/drtest/drtest.nerdygentleman.com/src/app.py 
  209  nvim /home/drtest/drtest.nerdygentleman.com/.env 
  210  service drtest status
  211  service drtest restart
  212  service drtest status
  213  nvim /home/drtest/drtest.nerdygentleman.com/.env 

replicator/app.py
import json
import logging
import logging.handlers
import os
import urllib
import random
import secrets
import string 
import sys
import itsdangerous
import requests
from dotenv import load_dotenv
import flask
from flask import Flask, request, Response

load_dotenv()
app = Flask(__name__)

# Configure logging
LOGGER = logging.getLogger()
log_formatter = logging.Formatter('%(name)s - %(levelname)s - %(message)s')

formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')

file_handler = logging.handlers.RotatingFileHandler(filename=os.environ['LOG_FILENAME'],
                                                    mode="a+",
                                                    maxBytes=50 * 1024 * 1024, # 50 MiB
                                                    backupCount=2,
                                                    encoding=None, delay=False)
file_handler.setFormatter(formatter)
file_handler.setLevel(logging.DEBUG)
LOGGER.setLevel(logging.DEBUG)
LOGGER.addHandler(file_handler)
LOGGER.info("GitHub Replicator app started")

AUTHORIZE_URI = "https://github.com/login/oauth/authorize"
ACCESS_TOKEN_URI = "https://github.com/login/oauth/access_token"
API_URI = "https://api.github.com"
SOURCE_REPO = "lkingsford/github_replicator"

@app.route('/')
def root():
    """Root that is called at / or index. This is generally the first page that
    a user will go to, and redirects them to sign in to GitHub"""
    redirect_params = {
        'client_id': os.environ['GITHUB_CLIENT_ID'],
        'redirect_uri': 'https://{host}/github_callback'.format(host=os.environ['HOST']),
        'scope': 'public_repo',
        'state': generate_state()
    }
    LOGGER.info("Sending redirect")
    LOGGER.debug("Using %s", redirect_params)
    params = urllib.parse.urlencode(redirect_params)
    redirect_uri = "{}?{}".format(AUTHORIZE_URI, params)
    LOGGER.debug("Redirect URI is %s", redirect_uri)
    return flask.redirect(redirect_uri)

@app.route('/github_callback')
def github_callback():
    """Route that is called by github, and gets an access token"""
    logging.info("Received Callback")
    logging.debug("Args: %s", request.args)
    # OAuth authorizes, and then redirects with a param with a code
    # that can be exchanged for an access token
    code = request.args.get('code')
    state = request.args.get('state')
    if not verify_state(state):
        return Response('State does not match - potential cross-site forgery',
                        401)
    authorize_params = {
        'client_id': os.environ['GITHUB_CLIENT_ID'],
        'client_secret': os.environ['GITHUB_CLIENT_SECRET'],
        'code': code
    }
    params = urllib.parse.urlencode(authorize_params)
    access_post_uri = "{}?{}".format(ACCESS_TOKEN_URI, params)
    logging.debug("Requesting token with '%s'", access_post_uri)
    response = requests.post(access_post_uri)
    # Response is in the form of 'access_token=<Access Token>&token_type=bearer
    # where <Access Token> is the actual token to use
    token_raw = response.content.decode('utf-8')
    logging.debug("Response was %s", token_raw)
    token = token_raw[token_raw.find('=') + 1 : token_raw.find('&')]
    logging.debug("Token is %s", token)

    # Duplicate repository
    # We're going to fork, so future users can get changes
    auth_headers = {"Authorization":"token {}".format(token)}
    logging.debug("Authorizing with %s", auth_headers)
    response = requests.post("{}/repos/{}/forks".format(API_URI, SOURCE_REPO),
                             headers=auth_headers)
    if response.status_code == 202:
        logging.info("Forked repo successfully. Response was %s",
            response.content.decode("utf-8"))
        # The repo is forked successfully. Redirect to it.
        json_response = json.loads(response.content)
        url = json_response["html_url"]
        return flask.redirect(url)
    else:
        logging.error("Failed to fork repo. Response was %s",
            response.content.decode("utf-8"))
        return "Failed to fork repo. Full response has been logged."

def generate_state():
    """Generate a state to pass to GitHub
    
    Normally, this might use a DB and store the actual state - but we're doing
    an encryption style state instead, so we don't have to keep additional
    infrastructure.
    
    We generate a random 12 character string, and then sign it with the
    state_key using itsdangerous. We use the same key to validate with
    itsdangerous the other way.
    """
    signer = itsdangerous.Signer(os.environ['STATE_KEY'])
    # Using secrets, as Python's doco makes clear that random is not suitable
    # for anything remotely security
    allowed_characters = string.ascii_letters + string.digits
    to_sign = ''.join([secrets.choice(allowed_characters) for i in range(12)])
    return signer.sign(to_sign).decode('utf-8')

def verify_state(state):
    """Verify the state GitHub returned is valid

    See `generate_state` for how this is verified
    """
    signer = itsdangerous.Signer(os.environ['STATE_KEY'])
    return signer.validate(state)

if __name__ == "__main__":
    app.run(debug=True, host='0.0.0.0')
